### PR TITLE
karmada operator: avoid invalid reconclie for karmada cr

### PR DESCRIPTION
**Pull request by calvin0327**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

I find that the karmada cr has been reconciled over and over again. 
Due to the two situation:
1. At the end of  every reconcile will update karmada cr status that will retrigger the next reconcile.
image

2. The status will update failed as the version conflicts.

**Which issue(s) this PR fixes**:
Fixes https://github.com/karmada-io/karmada/issues/3770

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by calvin0327** [comment]
I try to test it:
image

Create or delete the cr,  the log is printed only once that proof that it be reconciled once.
image

**Comment by calvin0327** [comment]
@Poor12 @junfsir @Fish-pro /cc

File: operator/pkg/controller/karmada/controller.go [hunk 1]
```diff
@@ -118,5 +129,25 @@ func (ctrl *Controller) ensureKarmada(ctx context.Context, karmada *operatorv1al
...
+		For(&operatorv1alpha1.Karmada{},
+			builder.WithPredicates(predicate.Funcs{
+				UpdateFunc: ctrl.onKarmadaUpdate,
+			})).
+		Complete(ctrl)
+}
+
+// onKarmadaUpdate returns a bool which represents whether the karmada event entry the queue.
+// In order to avoid invalid reconcile, we just focus on these two cases:
+// 1. when the karmada cr be deleted.
```
Comment (Poor12) [inline operator/pkg/controller/karmada/controller.go:142]:
```suggestion
// 1. when the karmada cr is deleted.
```

File: operator/pkg/controller/karmada/controller.go [hunk 2]
```diff
@@ -118,5 +129,25 @@ func (ctrl *Controller) ensureKarmada(ctx context.Context, karmada *operatorv1al
...
+// 2. when the spec of karmada cr is be updated.
```
Comment (Poor12) [inline operator/pkg/controller/karmada/controller.go:143]:
```suggestion
// 2. when the spec of karmada cr is updated.
```

File: operator/pkg/controller/karmada/controller.go [hunk 3]
```diff
@@ -118,5 +129,25 @@ func (ctrl *Controller) ensureKarmada(ctx context.Context, karmada *operatorv1al
...
-	return controllerruntime.NewControllerManagedBy(mgr).For(&operatorv1alpha1.Karmada{}).Complete(ctrl)
+	return controllerruntime.NewControllerManagedBy(mgr).
+		For(&operatorv1alpha1.Karmada{},
+			builder.WithPredicates(predicate.Funcs{
+				UpdateFunc: ctrl.onKarmadaUpdate,
+			})).
+		Complete(ctrl)
+}
+
+// onKarmadaUpdate returns a bool which represents whether the karmada event entry the queue.
```
Comment (Poor12) [inline operator/pkg/controller/karmada/controller.go:140]:
```suggestion
// onKarmadaUpdate returns a bool which represents whether the karmada events need to entry the queue.
```

**Review by Poor12 — COMMENTED** [review]
Just some typos.
/cc @Fish-pro @junfsir to help verify it.

**Comment by calvin0327** [comment]
updated.

**Review by RainbowMango — COMMENTED** [review]
/assign @Poor12

**Pull request merged**